src/material/core/datetime/date-selection.ts

```diff
@@ -0,0 +1,159 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {DateAdapter} from '@angular/material/core';
+import {Subject} from 'rxjs';
+
+export abstract class MatDateSelection<D> {
+  valueChanges = new Subject<void>();
+
+  constructor(protected readonly adapter: DateAdapter<D>) {}
+
+  dispose() {
+    this.valueChanges.complete();
+  }
+
+  abstract add(date: D | null): void;
+  abstract clone(): MatDateSelection<D>;
+  abstract getFirstSelectedDate(): D|null;
+  abstract getLastSelectedDate(): D|null;
+  abstract isComplete(): boolean;
+  abstract isSame(other: MatDateSelection<D>): boolean;
+  abstract isValid(): boolean;
+}
+
+export interface DateRange<D> {
+  start: D | null;
+  end: D | null;
+}
+
+/**
+ * Concrete implementation of a MatDateSelection that holds a single date.
+ */
+export class MatSingleDateSelection<D> extends MatDateSelection<D> {
+  private date: D | null = null;
+
+  constructor(adapter: DateAdapter<D>, date?: D | null) {
+    super(adapter);
+    if (date) {
+      this.date = date;
```

Comment:
If we did parameterize on `D` and `S` this could be moved into the base class, something to keep in mind